tensorboard/data/server/reservoir.rs

```diff
@@ -0,0 +1,329 @@
+    /// added to the reservoir. Usually called `N` in the literature.
+    seen: usize,
+}
+
+/// A `ReservoirControl` determines which records from a stream should be included into a
+/// reservoir, and which records they should evict. This is usually backed by a random number
+/// generator, but may be made deterministic for testing.
+pub trait ReservoirControl {
+    /// Upon seeing the latest record in a stream of `n` so far, rolls for the index into the
+    /// reservoir that should be evicted to make room for this record. If the result is greater
+    /// than or equal to the capacity of the reservoir, the record will be skipped; otherwise, it
+    /// will evict the record at the given index. The input `n` will be positive, and the result
+    /// must be in `0..n`.
+    ///
+    /// A useful implementation is to simply roll a random integer in `0..n`. Test code may wish to
+    /// provide an alternate, deterministic implementation.
+    fn destination(&mut self, n: usize) -> usize;
+}
+
+impl<R: Rng> ReservoirControl for R {
+    fn destination(&mut self, n: usize) -> usize {
+        Uniform::from(0..n).sample(self)
+    }
+}
+
+impl<T> StageReservoir<T, ChaCha20Rng> {
+    /// Creates a new reservoir with the specified capacity, using a fixed-seed random number
+    /// generator for reservoir control. All reservoirs created by this function will use the same
+    /// sequence of random numbers.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn new(capacity: usize) -> Self {
+        Self::with_control(capacity, ChaCha20Rng::seed_from_u64(0))
+    }
+}
+
+impl<T, C: ReservoirControl> StageReservoir<T, C> {
+    /// Offers a record to the reservoir. The reservoir will always include the latest record.
+    /// Other than the latest record, the records kept form a simple random sample of the stream
+    /// (or at least approximately so in the case of preemptions).
+    pub fn offer(&mut self, step: Step, v: T) {
+        self.seen += 1;
+        let dst = self.ctl.destination(self.seen);
+
+        // Didn't make the cut? Keep-last only.
+        if dst >= self.capacity {
+            if self.staged_items.pop().is_none() {
+                self.committed_steps.pop();
+            }
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Room to spare? Add to end.
+        let num_stored = self.committed_steps.len() + self.staged_items.len();
+        if num_stored < self.capacity {
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Otherwise, we're evicting.
+        if dst < self.committed_steps.len() {
+            self.committed_steps.remove(dst);
+        } else {
+            self.staged_items.remove(dst - self.committed_steps.len());
+        }
+        self.staged_items.push((step, v))
```

Comment:
This confused me for a sec since the lack of trailing semicolon made me wonder if this was the return value.  (Which I guess it technically is, but not in a way that's communicating anything useful.)  Is that idiomatic?  It seems a bit clearer to me to always include a trailing semicolon if the final expression is not being used as a return value.